README.md
# Features Maximization Metric

[![ci](https://github.com/cognitivefactory/features-maximization-metric/workflows/ci/badge.svg)](https://github.com/cognitivefactory/features-maximization-metric/actions?query=workflow%3Aci)
[![documentation](https://img.shields.io/badge/docs-mkdocs%20material-blue.svg?style=flat)](https://cognitivefactory.github.io/features-maximization-metric/)
[![pypi version](https://img.shields.io/pypi/v/cognitivefactory-features-maximization-metric.svg)](https://pypi.org/project/cognitivefactory-features-maximization-metric/)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7646382.svg)](https://doi.org/10.5281/zenodo.7646382)

Implementation of _Features Maximization Metric_, an unbiased metric aimed at estimate the quality of an unsupervised classification.


## <a name="Description"></a> Quick description

_Features Maximization_ (`FMC`) is a features selection method described in `Lamirel, J.-C., Cuxac, P., & Hajlaoui, K. (2016). A Novel Approach to Feature Selection Based on Quality Estimation Metrics. In Advances in Knowledge Discovery and Management (pp. 121–140). Springer International Publishing. https://doi.org/10.1007/978-3-319-45763-5_7`.

This metric is computed by applying the following steps:

1. Compute the ***Features F-Measure*** metric (based on ***Features Recall*** and ***Features Predominance*** metrics).

    > (a) The ***Features Recall*** `FR[f][c]` for a given class `c` and a given feature `f` is the ratio between
    > the sum of the vectors weights of the feature `f` for data in class `c`
    > and the sum of all vectors weights of feature `f` for all data.
    > It answers the question: "_Can the feature `f` distinguish the class `c` from other classes `c'` ?_"

    > (b) The ***Features Predominance*** `FP[f][c]` for a given class `c` and a given feature `f` is the ratio between
    > the sum of the vectors weights of the feature `f` for data in class `c`
    > and the sum of all vectors weights of all feature `f'` for data in class `c`.
    > It answers the question: "_Can the feature `f` better identify the class `c` than the other features `f'` ?_"

    > (c) The ***Features F-Measure*** `FM[f][c]` for a given class `c` and a given feature `f` is
    > the harmonic mean of the ***Features Recall*** (a) and the ***Features Predominance*** (c).
    > It answers the question: "_How much information does the feature `f` contain about the class `c` ?_"

2. Compute the ***Features Selection*** (based on ***F-Measure Overall Average*** comparison).

    > (d) The ***F-Measure Overall Average*** is the average of ***Features F-Measure*** (c) for all classes `c` and for all features `f`.
    > It answers the question: "_What are the mean of information contained by features in all classes ?_"

    > (e) A feature `f` is ***Selected*** if and only if it exist at least one class `c` for which the ***Features F-Measure*** (c) `FM[f][c]` is bigger than the ***F-Measure Overall Average*** (d).
    > It answers the question: "_What are the features which contain more information than the mean of information in the dataset ?_"

    > (f) A Feature `f` is ***Deleted*** if and only if the ***Features F-Measure*** (c) `FM[f][c]` is always lower than the ***F-Measure Overall Average*** (d) for each class `c`.
    > It answers the question: "_What are the features which do not contain more information than the mean of information in the dataset ?_"

3. Compute the ***Features Contrast*** and ***Features Activation*** (based on ***F-Measure Marginal Averages*** comparison).

    > (g) The ***F-Measure Marginal Averages*** for a given feature `f` is the average of ***Features F-Measure*** (c) for all classes `c` and for the given feature `f`.
    > It answers the question: "_What are the mean of information contained by the feature `f` in all classes ?_"

    > (h) The ***Features Contrast*** `FC[f][c]` for a given class `c` and a given selected feature `f` is the ratio between
    > the ***Features F-Measure*** (c) `FM[f][c]`
    > and the ***F-Measure Marginal Averages*** (g) for selected feature f
    > put to the power of an ***Amplification Factor***.
    > It answers the question: "_How relevant is the feature `f` to distinguish the class `c` ?_"

    > (i) A selected Feature `f` is ***Active*** for a given class `c` if and only if the ***Features Contrast*** (h) `FC[f][c]` is bigger than `1.0`.
    > It answers the question : "_For which classes a selected feature `f` is relevant ?_"

This metric is an **efficient method** to:

- **identify relevant features** of a dataset modelization;
- **describe association** between vectors features and data classes;
- **increase contrast** between data classes.


## <a name="Documentation"></a> Documentation

- [Main documentation](https://cognitivefactory.github.io/features-maximization-metric/)


## <a name="Installation"></a> Installation

Features Maximization Metric requires [`Python`](https://www.python.org/) 3.8 or above.

To install with [`pip`](https://github.com/pypa/pip):

```bash
# install package
python3 -m pip install cognitivefactory-features-maximization-metric
```

To install with [`pipx`](https://github.com/pypa/pipx):

```bash
# install pipx
python3 -m pip install --user pipx

# install package
pipx install --python python3 cognitivefactory-features-maximization-metric
```


## <a name="Development"></a> Development

To work on this project or contribute to it, please read:

- the [Copier PDM](https://pawamoy.github.io/copier-pdm/) template documentation ;
- the [Contributing](https://cognitivefactory.github.io/features-maximization-metric/contributing/) page for environment setup and development help ;
- the [Code of Conduct](https://cognitivefactory.github.io/features-maximization-metric/code_of_conduct/) page for contribution rules.


## <a name="References"></a> References

- **Features Maximization Metric**: `Lamirel, J.-C., Cuxac, P., & Hajlaoui, K. (2016). A Novel Approach to Feature Selection Based on Quality Estimation Metrics. In Advances in Knowledge Discovery and Management (pp. 121–140). Springer International Publishing. https://doi.org/10.1007/978-3-319-45763-5_7`
- **V-Measure**: `Rosenberg, Andrew & Hirschberg, Julia. (2007). V-Measure: A Conditional Entropy-Based External Cluster Evaluation Measure. 410-420.`


## <a name="How to cite"></a> How to cite	

`Schild, E. (2023). cognitivefactory/features-maximization-metric. Zenodo. https://doi.org/10.5281/zenodo.7646382.`

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->
## [1.0.0](https://github.com/cognitivefactory/features-maximization-metric/releases/tag/1.0.0) - 2023-11-14

<small>[Compare with 0.1.1](https://github.com/cognitivefactory/features-maximization-metric/compare/0.1.1...1.0.0)</small>

### Build
- prepare production release ([7688333](https://github.com/cognitivefactory/features-maximization-metric/commit/76883335c6aeb99639b5bb638e11140dbf0a7566) by SCHILD Erwan).

### Code Refactoring
- update comment ([b106764](https://github.com/cognitivefactory/features-maximization-metric/commit/b106764ef83b3ae1e24680dee91c275b23eb8ee4) by SCHILD Erwan).


## [0.1.1](https://github.com/cognitivefactory/features-maximization-metric/releases/tag/0.1.1) - 2023-02-16

<small>[Compare with 0.1.0](https://github.com/cognitivefactory/features-maximization-metric/compare/0.1.0...0.1.1)</small>

### Build
- fix git-changelog dependancy ([a6b6684](https://github.com/cognitivefactory/features-maximization-metric/commit/a6b66843e4de6d93a2c910e38b677881c86dab6b) by SCHILD Erwan).

## [0.1.0](https://github.com/cognitivefactory/features-maximization-metric/releases/tag/0.1.0) - 2023-02-16

<small>[Compare with first commit](https://github.com/cognitivefactory/features-maximization-metric/compare/0bfc7fe5356231f075af47c7a8113db8f32ba9f0...0.1.0)</small>

### Build
- configure project template ([741b82c](https://github.com/cognitivefactory/features-maximization-metric/commit/741b82cb58efef912982895d83fb483a6d7adc00) by SCHILD Erwan).
- add .gitignore ([9e57355](https://github.com/cognitivefactory/features-maximization-metric/commit/9e57355197c5278beb3155cc75b6ac6a65579360) by SCHILD Erwan).

### Features
- implement Feature Maximization Metric ([44554c3](https://github.com/cognitivefactory/features-maximization-metric/commit/44554c3ee6bb79193c91407e44d425769d01582a) by SCHILD Erwan).


tests/__init__.py
"""Tests suite for `r_wnlp.features_maximization`."""

from pathlib import Path

TESTS_DIR = Path(__file__).parent
TMP_DIR = TESTS_DIR / "tmp"
FIXTURES_DIR = TESTS_DIR / "fixtures"


scripts/multirun.sh
#!/usr/bin/env bash
set -e

# un/comment this if you want to enable/disable multiple Python versions
#### PYTHON_VERSIONS="${PYTHON_VERSIONS-3.8 3.9 3.10 3.11}"

restore_previous_python_version() {
    if pdm use -f "$1" &>/dev/null; then
        echo "> Restored previous Python version: ${1##*/}"
    fi
}

if [ -n "${PYTHON_VERSIONS}" ]; then
    old_python_version="$(pdm config python.path)"
    echo "> Currently selected Python version: ${old_python_version##*/}"
    trap "restore_previous_python_version ${old_python_version}" EXIT
    for python_version in ${PYTHON_VERSIONS}; do
        if pdm use -f "${python_version}" &>/dev/null; then
        echo "> pdm run $@ (${python_version})"
            pdm run "$@"
        else
            echo "> pdm use -f ${python_version}: Python interpreter not available?" >&2
        fi
    done
else
    pdm run "$@"
fi


scripts/setup.sh
#!/usr/bin/env bash
set -e

# un/comment this if you want to enable/disable multiple Python versions
#### PYTHON_VERSIONS="${PYTHON_VERSIONS-3.8 3.9 3.10 3.11}"

install_with_pipx() {
    if ! command -v "$1" &>/dev/null; then
        if ! command -v pipx &>/dev/null; then
            python3 -m pip install --user pipx
        fi
        pipx install "$1"
    fi
}

install_with_pipx pdm

restore_previous_python_version() {
    if pdm use -f "$1" &>/dev/null; then
        echo "> Restored previous Python version: ${1##*/}"
    fi
}

if [ -n "${PYTHON_VERSIONS}" ]; then
    if old_python_version="$(pdm config python.path 2>/dev/null)"; then
        echo "> Currently selected Python version: ${old_python_version##*/}"
        trap "restore_previous_python_version ${old_python_version}" EXIT
    fi
    for python_version in ${PYTHON_VERSIONS}; do
        if pdm use -f "${python_version}" &>/dev/null; then
            echo "> Using Python ${python_version} interpreter"
            pdm install
        else
            echo "> pdm use -f ${python_version}: Python interpreter not available?" >&2
        fi
    done
else
    pdm install
fi
